File: pyci/src/cepa0.cpp
/* This file is part of PyCI.
 *
 * PyCI is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * PyCI is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with PyCI. If not, see <http://www.gnu.org/licenses/>. */

#include <pyci.h>

namespace pyci {

void perform_op_cepa0_thread(const double *data, const long *indptr, const long *indices,
                             const double *x, double *y, const long start, const long end,
                             const long refind, const double h_refind) {
    long j, jstart, jend = indptr[start];
    double val;
    for (long i = start; i < end; ++i) {
        jstart = jend;
        jend = indptr[i + 1];
        val = 0.0;
        if (i == refind) {
            for (j = jstart; j < jend; ++j) {
                if (indices[j] == refind)
                    val -= x[indices[j]];
                else
                    val += data[j] * x[indices[j]];
            }
        } else {
            for (j = jstart; j < jend; ++j) {
                if (indices[j] == i)
                    val += (data[j] - h_refind) * x[indices[j]];
                else if (indices[j] != refind)
                    val += data[j] * x[indices[j]];
            }
        }
        y[i] = -val;
    }
}

void SparseOp::perform_op_cepa0(const double *x, double *y, const long refind) const {
    long nthread = get_num_threads(), start, end;
    long chunksize = nrow / nthread + ((nrow % nthread) ? 1 : 0);
    double h_refind = get_element(refind, refind);
    Vector<std::thread> v_threads;
    v_threads.reserve(nthread);
    for (long i = 0; i < nthread; ++i) {
        start = i * chunksize;
        end = (start + chunksize < nrow) ? start + chunksize : nrow;
        v_threads.emplace_back(&perform_op_cepa0_thread, &data[0], &indptr[0], &indices[0], x, y,
                               start, end, refind, h_refind);
    }
    for (auto &thread : v_threads)
        thread.join();
}

void SparseOp::perform_op_transpose_cepa0(const double *x, double *y, const long refind) const {
    long i;
    for (i = 0; i < ncol; ++i)
        y[i] = 0;
    long j, jstart, jend = indptr[0];
    double h_refind = get_element(refind, refind);
    for (i = 0; i < nrow; ++i) {
        jstart = jend;
        jend = indptr[i + 1];
        for (j = jstart; j < jend; ++j) {
            if (indices[j] == i)
                y[indices[j]] += (h_refind - data[j]) * x[i];
            else
                y[indices[j]] -= data[j] * x[i];
        }
    }
    y[refind] = x[refind];
}

void SparseOp::rhs_cepa0(double *b, const long refind) const {
    long iend = indptr[refind + 1];
    for (long i = indptr[refind]; i != iend; ++i) {
        if (indices[i] >= nrow)
            break;
        b[indices[i]] = data[i];
    }
}

Array<double> SparseOp::py_matvec_cepa0(const Array<double> x, const long refind) const {
    if (symmetric)
        throw std::runtime_error("cannot run CEPA0 with sparse_op(symmetric=True)");
    Array<double> y(nrow);
    perform_op_cepa0(reinterpret_cast<const double *>(x.request().ptr),
                     reinterpret_cast<double *>(y.request().ptr), refind);
    return y;
}

Array<double> SparseOp::py_rmatvec_cepa0(Array<double> x, const long refind) const {
    if (symmetric)
        throw std::runtime_error("cannot run CEPA0 with sparse_op(symmetric=True)");
    Array<double> y(ncol);
    perform_op_transpose_cepa0(reinterpret_cast<const double *>(x.request().ptr),
                               reinterpret_cast<double *>(y.request().ptr), refind);
    return y;
}

Array<double> SparseOp::py_rhs_cepa0(const long refind) const {
    if (symmetric)
        throw std::runtime_error("cannot run CEPA0 with sparse_op(symmetric=True)");
    Array<double> y(nrow);
    rhs_cepa0(reinterpret_cast<double *>(y.request().ptr), refind);
    return y;
}

} // namespace pyci

File: pyci/cepa0.py
# This file is part of PyCI.
#
# PyCI is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your
# option) any later version.
#
# PyCI is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License
# along with PyCI. If not, see <http://www.gnu.org/licenses/>.

r"""PyCI CEPA0 module."""

from typing import Union, Tuple

import numpy as np

from scipy.sparse.linalg import lgmres, lsqr

from . import pyci


__all__ = [
    "CEPA0",
]


class CEPA0:
    r"""CEPA0 solver class."""

    @property
    def shape(self) -> Tuple[int, int]:
        r"""
        Shape of the linear operator.

        Returns
        -------
        nrow : int
            Number of rows.
        ncol : int
            Number of columns.

        """
        return self.op.shape

    @property
    def dtype(self) -> np.dtype:
        r"""
        Data type of the linear operator.

        Returns
        -------
        dtype : numpy.dtype
            Data type of the linear operator.

        """
        return self.op.dtype

    def __init__(
        self,
        *args: Union[Tuple[pyci.wavefunction, pyci.hamiltonian], pyci.sparse_op],
        ref: int = 0,
        damp: float = np.inf,
    ) -> None:
        r"""
        Initialize the CEPA0 solver.

        Parameters
        ----------
        args : (pyci.sparse_op,) or (pyci.hamiltonian, pyci.wavefunction)
            System to solve.
        ref : int, default=0
            Index of reference determinant.
        damp : float, default=numpy.inf
            Damping parameter.

        """
        if len(args) == 1:
            self.op = args[0]
            if self.op.symmetric:
                raise TypeError("Cannot run CEPA0 on symmetric matrix operator")
            elif self.op.shape[0] < self.op.shape[1]:
                raise ValueError("Cannot solve underdetermined system (nrow < ncol)")
        elif len(args) == 2:
            self.op = pyci.sparse_op(*args, symmetric=False)
        else:
            raise ValueError("must pass `ham, wfn` or `op`")
        self.ref = ref
        self.h_ref = self.op.get_element(ref, ref)
        self.damp = -1.0 / damp ** 2

    def __call__(self, x: np.ndarray) -> np.ndarray:
        r"""
        Apply the linear operator to a vector ``x``.

        Parameters
        ----------
        x : numpy.ndarray
            Vector to which linear operator will be applied.

        Returns
        -------
        y : numpy.ndarray
            Result of applying linear operator to ``x``.

        """
        return self.matvec(x)

    def matvec(self, x: np.ndarray) -> np.ndarray:
        r"""
        Apply the linear operator to a vector ``x``.

        Parameters
        ----------
        x : numpy.ndarray
            Vector to which linear operator will be applied.

        Returns
        -------
        y : numpy.ndarray
            Result of applying linear operator to ``x``.

        """
        op = self.op
        ref = self.ref
        h_ref = self.h_ref
        damp = self.damp
        y = np.zeros(op.shape[0], dtype=pyci.c_double)
        for i in range(op.shape[0]):
            h_ref_damp = np.exp(damp * x[i] ** 2) * h_ref
            if i == ref:
                for j in range(self.op.shape[1]):
                    if j == ref:
                        y[i] += x[j]
                    else:
                        y[i] -= op.get_element(i, j) * x[j]
            else:
                for j in range(self.op.shape[1]):
                    if i == j:
                        y[i] += (h_ref_damp - op.get_element(i, j)) * x[j]
                    elif j != ref:
                        y[i] -= op.get_element(i, j) * x[j]
        return y

    def rmatvec(self, x: np.ndarray) -> np.ndarray:
        r"""
        Apply the transpose of the linear operator to a vector ``x``.

        Parameters
        ----------
        x : numpy.ndarray
            Vector to which linear operator will be applied.

        Returns
        -------
        y : numpy.ndarray
            Result of applying linear operator to ``x``.

        """
        op = self.op
        h_ref = self.h_ref
        damp = self.damp
        y = np.zeros(op.shape[1], dtype=pyci.c_double)
        for i in range(op.shape[0]):
            h_ref_damp = np.exp(damp * x[i] ** 2) * h_ref
            for j in range(self.op.shape[1]):
                if j == i:
                    y[j] += (h_ref_damp - op.get_element(i, j)) * x[i]
                else:
                    y[j] -= op.get_element(i, j) * x[i]
        y[self.ref] = x[self.ref]
        return y

    def rhs(self, x: np.ndarray) -> np.ndarray:
        r"""
        Construct the right-hand side vector for the linear system of equations.

        Parameters
        ----------
        x : numpy.ndarray
            Initial guess.

        Returns
        -------
        rhs : numpy.ndarray
            Right-hand side vector.

        """
        op = self.op
        ref = self.ref
        y = np.zeros(op.shape[0], dtype=pyci.c_double)
        for i in range(op.shape[0]):
            y[i] = op.get_element(i, ref)
        y -= self.matvec(x)
        return y

    def solve(
        self, c0: np.ndarray = None, e0: float = None, maxiter: int = 1000, tol: float = 1.0e-12,
    ) -> Tuple[np.ndarray, np.ndarray]:
        r"""
        Solve the CEPA0 problem.

        Parameters
        ----------
        c0 : numpy.ndarray, optional
            Initial guess for lowest eigenvector.
        e0 : float, optional
            Initial guess for energy.
        maxiter : int, default=1000
            Maximum number of iterations to perform.
        tol : float, default=1.0e-12
            Convergence tolerance.

        Returns
        -------
        es : numpy.ndarray
            Energy.
        cs : numpy.ndarray
            Coefficient vector.

        """
        c0 = np.zeros(self.op.shape[1], dtype=pyci.c_double) if c0 is None else c0 / c0[self.ref]
        c0[self.ref] = self.op.get_element(self.ref, self.ref) if e0 is None else e0 - self.op.ecore
        if self.op.shape[0] == self.op.shape[1]:
            cs = lgmres(self, self.rhs(c0), maxiter=maxiter, tol=tol, atol=tol)[0]
        else:
            cs = lsqr(self, self.rhs(c0), maxiter=maxiter, atol=tol, btol=tol)[0]
        cs += c0
        es = np.full(1, cs[self.ref] + self.op.ecore, dtype=pyci.c_double)
        cs[self.ref] = 1
        return es, cs[None, :]
